File: Scala/src/main/scala/TextbookTrack/Chapter05/BA5C.scala
package TextbookTrack.Chapter05

object BA5C {
  import scala.annotation.tailrec

  private def collectLCS(backtrack: Array[Array[Byte]], string: String, n1: Int, n2: Int): String = {
    @tailrec
    def loop(lcs: List[Char], ix: Int, jy: Int): String = {
      if (ix == 0 || jy == 0) lcs.mkString
      else if (backtrack(ix - 1)(jy - 1) == 0) loop(string(ix - 1) :: lcs, ix - 1, jy - 1)
      else if (backtrack(ix - 1)(jy - 1) == -1) loop(lcs, ix - 1, jy)
      else loop(lcs, ix, jy - 1)
    }

    loop(Nil, n1, n2)
  }

  def calcLongestCommonSubsequence(string1: String, string2: String): String = {
    val n1: Int = string1.length
    val n2: Int = string2.length
    val table: Array[Array[Int]] = Array.fill(n1 + 1, n2 + 1)(0)
    val backtrack: Array[Array[Byte]] = Array.fill(n1, n2)(0)

    for {
      (c1, ix) <- string1.zipWithIndex
      (c2, jy) <- string2.zipWithIndex
    } {
      val pathDown: Int = table(ix)(jy + 1)
      val pathRight: Int = table(ix + 1)(jy)
      val pathDiag: Int = table(ix)(jy) + (if (c1 == c2) 1 else 0)
      val maxPath: Int = List(pathDown, pathRight, pathDiag).max
      table(ix + 1)(jy + 1) = maxPath
      backtrack(ix)(jy) = if (maxPath == pathDown) -1 else if (maxPath == pathRight) 1 else 0
    }

    collectLCS(backtrack, string1, n1, n2)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val string1: String = reader.next()
    val string2: String = reader.next()
    val result: String = calcLongestCommonSubsequence(string1, string2)
    println(result)
  }
}

File: Python/textbook_track/chapter05/ba5c.py
# Find a Longest Common Subsequence of Two Strings
import sys


def collect_longest_common_subsequence(backtrack, string, ix, jy):
    lcs = []
    while ix > 0 and jy > 0:
        if backtrack[ix-1][jy-1] == 0:
            lcs.append(string[ix-1])
            ix -= 1
            jy -= 1
        elif backtrack[ix-1][jy-1] == -1:
            ix -= 1
        else:
            jy -= 1
    return ''.join(lcs[::-1])


def calc_longest_common_subsequence(string1, string2):
    n1 = len(string1)
    n2 = len(string2)
    table = [[0] * (n2+1) for _ in range(n1+1)]
    backtrack = [[0] * n2 for _ in range(n1)]

    for ix, c1 in enumerate(string1):
        for jy , c2 in enumerate(string2):
            path_down = table[ix][jy+1]
            path_right = table[ix+1][jy]
            path_diag = table[ix][jy] + int(c1 == c2)
            max_path = max(path_down, path_right, path_diag)
            table[ix+1][jy+1] = max_path

            if max_path == path_down:
                backtrack[ix][jy] = -1
            elif max_path == path_right:
                backtrack[ix][jy] = 1

    return collect_longest_common_subsequence(backtrack, string1, n1, n2)


def main():
    data = sys.stdin.read().splitlines()
    string1 = data[0]
    string2 = data[1]
    result = calc_longest_common_subsequence(string1, string2)
    print(result)


if __name__ == '__main__':
    main()
